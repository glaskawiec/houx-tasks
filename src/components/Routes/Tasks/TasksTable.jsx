import React from 'react';
import moment from 'moment';
import { useHoux } from 'houx';
import RemoveButton from './RemoveButton';
import CompleteButton from './CompleteButton';

export default function TasksTable() {
  const [state] = useHoux();

  return (
    <table className="table table-dark">
      <thead>
        <tr>
          <th>#Id</th>
          <th>Priority</th>
          <th>Text</th>
          <th>Date</th>
          <th>Actions</th>
        </tr>
      </thead>
      <tbody>
        {state.tasks.tasks.map((task) => (
          <tr key={task.id}>
            <td>{task.id}</td>
            <td>{task.priority}</td>
            <td>{task.name}</td>
            <td>{moment(task.date).format('MMMM Do YYYY, h:mm a').toString()}</td>
            <td>
              <CompleteButton task={task} style={{ marginRight: '1em' }} />
              <RemoveButton itemId={task.id} />
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}