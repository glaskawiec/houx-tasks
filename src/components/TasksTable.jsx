import React, { useState } from 'react';
import { useHoux } from 'houx';
import { Table } from 'react-bootstrap';
import styled from 'styled-components';
import RemoveButton from './RemoveButton';
import EditButton from './EditButton';

const CollapseButton = styled.span`
  display: inline-block;
  padding: 0.2rem;
  background-color: transparent;
  color: white;
  cursor: pointer;
  border: 0;
`;

const Text = styled.p`
  display: inline-block;
`;

const DescriptionCell = styled.td`
  max-width: 10rem;
`;

const Row = styled.tr`
 background-color: ${(props) => props.backgroundColor};
 color: ${(props) => props.color};
`;

const TasksTable = () => {
  const { state } = useHoux();
  const [isCollapsed, setIsCollapsed] = useState({});

  const onCollapseClick = (task) => setIsCollapsed({ ...isCollapsed, [task.id]: !isCollapsed[task.id] });

  const headers = ['Id', 'Title', 'Description', 'Actions'];

  return (
    <Table bordered variant="dark" responsive>
      <thead>
        <tr>
          {headers.map((header) => <th key={header}>{header}</th>)}
        </tr>
      </thead>
      <tbody>
        {state.tasks.tasks
          .sort((a, b) => a.id - b.id)
          .map((task) => (
            <Row
              color={task.colors.text}
              backgroundColor={task.colors.background}
              data-testid={`table-row-${task.id}`}
              key={task.id}
            >
              <td data-testid={`table-id-cell-${task.id}`}>
                <Text>
                  {task.id}
                </Text>
              </td>
              <td data-testid={`table-title-cell-${task.id}`}>
                <Text>
                  {task.title}
                </Text>
              </td>
              <DescriptionCell data-testid={`table-description-cell-${task.id}`}>
                {!isCollapsed[task.id] ? <></> : (
                  <Text>
                    {task.description}
                  </Text>
                )}
                {task.description && task.description.length > 0
                  ? (
                    <CollapseButton
                      data-testid={`collapse-button-task-${task.id}`}
                      onClick={() => onCollapseClick(task)}
                    >
                      {isCollapsed[task.id] ? '️⬅️' : '➡'}
                    </CollapseButton>
                  ) : null}
              </DescriptionCell>
              <td>
                <RemoveButton id={task.id} />
                <EditButton id={task.id} />
              </td>
            </Row>
          ))}
      </tbody>
    </Table>
  );
};

export default TasksTable;
