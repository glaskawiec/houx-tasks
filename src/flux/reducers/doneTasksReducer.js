import { DONE_TASKS_ADD, DONE_TASKS_REMOVE } from "../actionTypes/doneTasksActionTypes";

export const initialState = [];

const doneTasksReducer = (state = initialState, action) => {
    switch (action && action.type) {
        case DONE_TASKS_ADD:
            return [...state, action.task]
        case DONE_TASKS_REMOVE:
            return state.filter((task) => task.id !== action.id)
        default: return state;
    }
}

export default doneTasksReducer;