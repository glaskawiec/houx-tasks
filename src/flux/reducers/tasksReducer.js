import {
  TASKS_ADD,
  TASKS_REMOVE,
  TASKS_COMPLETE,
  TASKS_EDIT,
  TASKS_OPEN_ADD_MODAL,
  TASKS_OPEN_EDIT_MODAL,
  TASKS_CLOSE_MODAL, TASKS_CHANGE_ORDER,
} from '../actionTypes/tasksActionTypes';

export const initialState = {
  tasks: [
    {
      id: 0,
      title: 'Example task',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
      colors: {
        text: 'pink',
      },
    },
    {
      id: 1,
      title: 'Example task 2',
      description: '',
      colors: {
        text: 'purple',
      },
    },
    {
      id: 2,
      title: 'Example task 3',
      description: '',
      colors: {
        text: 'yellow',
      },
    },
  ],
  nextId: 3,
  isEditModalOpen: false,
  currentEditTaskId: null,
  isAddModalOpen: false,
};

const tasksReducer = (state = initialState, action) => {
  switch (action && action.type) {
    case TASKS_ADD: {
      const newTodo = {
        ...action.task,
        id: state.nextId,
      };

      return {
        ...state,
        nextId: state.nextId + 1,
        tasks: [...state.tasks, newTodo],
      };
    }
    case TASKS_REMOVE: {
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.id),
      };
    }
    case TASKS_COMPLETE:
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.id),
      };
    case TASKS_OPEN_ADD_MODAL:
      return {
        ...state,
        isAddModalOpen: true,

      };
    case TASKS_CLOSE_MODAL:
      return {
        ...state,
        isEditModalOpen: false,
        isAddModalOpen: false,
      };
    case TASKS_OPEN_EDIT_MODAL:
      return {
        ...state,
        isEditModalOpen: true,
        currentEditTaskId: action.id,
      };
    case TASKS_EDIT: {
      const { currentEditTaskId } = state;
      return {
        ...state,
        tasks: [
          ...state.tasks.filter((task) => task.id !== action.id),
          { ...action.editedTask, id: currentEditTaskId },
        ],
        currentEditTaskId: null,
      };
    }
    case TASKS_CHANGE_ORDER: {
      const {orderDown, id} = action;
      const tasksCopy = [...state.tasks];
      let tasksToOrderIndex = 0;

      for( ; tasksToOrderIndex < tasksCopy.length; ++tasksToOrderIndex){
        if(tasksCopy[tasksToOrderIndex].id === id){
          break;
        }
      }

      if(orderDown){
        const previousId = tasksCopy[tasksToOrderIndex+1].id;
        tasksCopy[tasksToOrderIndex].id = previousId;
        tasksCopy[tasksToOrderIndex+1].id = id;
      } else {
        const previousId = tasksCopy[tasksToOrderIndex-1].id;
        tasksCopy[tasksToOrderIndex].id = previousId;
        tasksCopy[tasksToOrderIndex-1].id = id;
      }

      return {
        ...state,
        tasks: tasksCopy
      };
    }
    default: return state;
  }
};

export default tasksReducer;
