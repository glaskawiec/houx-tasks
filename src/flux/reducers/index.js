import tasksReducer from './tasksReducer';

const reducers = {
  tasks: tasksReducer,
};

export default reducers;
