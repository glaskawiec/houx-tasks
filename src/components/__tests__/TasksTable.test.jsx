import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { getNodeText } from '@testing-library/dom';
import { HouxProvider } from 'houx';
import TasksTable from '../TasksTable';
import reducers from '../../flux/reducers';
import tasksReducer from '../../flux/reducers/tasksReducer';

describe('TasksTable component unit tests', () => {
  const mockedState = {
    tasks: [
      {
        id: 0,
        title: 'Test task 1',
        description: 'Test description 1',
        colors: {
          text: 'pink',
          background: 'yellow',
        },
      },
      {
        id: 1,
        title: 'Test task 2',
        description: '',
        colors: {
          text: 'purple',
        },
      },
    ],
    nextId: 2,
    currentEditTaskId: null,
    isEditModalOpen: false,
    isAddModalOpen: false,
  };

  const mockedReducers = {
    tasks: (state, action) => tasksReducer(mockedState, action),
  };

  test('Header should render properly', () => {
    const { container } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('Clicking on remove task button should remove row properly row from table', () => {
    const { queryByTestId } = render(
      <HouxProvider reducers={reducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(queryByTestId('table-row-0')).toBeInTheDocument();

    fireEvent.click(queryByTestId('remove-task-0-button'));

    expect(queryByTestId('table-row-0')).not.toBeInTheDocument();
  });

  test('Should id be shown properly', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getNodeText(getByTestId('table-id-cell-0').firstChild)).toBe(mockedState.tasks[0].id.toString());
    expect(getNodeText(getByTestId('table-id-cell-1').firstChild)).toBe(mockedState.tasks[1].id.toString());
  });

  test('Should title be shown properly', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getNodeText(getByTestId('table-title-cell-0').firstChild)).toBe(mockedState.tasks[0].title.toString());
    expect(getNodeText(getByTestId('table-title-cell-1').firstChild)).toBe(mockedState.tasks[1].title.toString());
  });

  test('Should description cell be empty if there is no description', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getByTestId('table-description-cell-1').firstChild).toBe(null);
  });

  test('Should description be collapsed by default if there is description', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getNodeText(getByTestId('table-description-cell-0').firstChild)).toBe('➡');
  });

  test('Should description be shown after clicking on collapse button', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getNodeText(getByTestId('table-description-cell-0').firstChild)).toBe('➡');

    fireEvent.click(getByTestId('collapse-button-task-0'));

    expect(getNodeText(getByTestId('table-description-cell-0').firstChild)).toBe(mockedState.tasks[0].description);
  });

  test('Should task row have possibility to add custom background color', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getByTestId('table-row-0')).toHaveStyle(`background-color: ${mockedState.tasks[0].colors.background}`);
  });

  test('Should task row have possibility to add custom text color', () => {
    const { getByTestId } = render(
      <HouxProvider reducers={mockedReducers}>
        <TasksTable />
      </HouxProvider>,
    );

    expect(getByTestId('table-row-0')).toHaveStyle(`color: ${mockedState.tasks[0].colors.text}`);
  });

  test('User should be able to order task to 1 item down', () => {
    const { getByTestId } = render(
        <HouxProvider reducers={reducers}>
          <TasksTable />
        </HouxProvider>,
    );


    expect(getNodeText(getByTestId('table-title-cell-0').firstChild)).toBe('Example task');

    fireEvent.click(getByTestId('order-down-task-0-button'));

    expect(getNodeText(getByTestId('table-title-cell-1').firstChild)).toBe('Example task');
  });

  test('User should be able to order task to 1 item up', () => {
    const { getByTestId } = render(
        <HouxProvider reducers={reducers}>
          <TasksTable />
        </HouxProvider>,
    );


    expect(getNodeText(getByTestId('table-title-cell-2').firstChild)).toBe('Example task 2');

    fireEvent.click(getByTestId('order-up-task-2-button'));

    expect(getNodeText(getByTestId('table-title-cell-1').firstChild)).toBe('Example task 2');
  });
});
