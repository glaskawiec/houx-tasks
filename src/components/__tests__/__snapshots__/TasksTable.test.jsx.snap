// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksTable component unit tests Header should render properly 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-dark table-bordered"
    >
      <thead>
        <tr>
          <th>
            Id
          </th>
          <th>
            Title
          </th>
          <th>
            Description
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="sc-AxhCb kxHtPP"
          color="pink"
          data-testid="table-row-0"
        >
          <td
            data-testid="table-id-cell-0"
          >
            <p
              class="sc-AxirZ iivBpk"
            >
              0
            </p>
          </td>
          <td
            data-testid="table-title-cell-0"
          >
            <p
              class="sc-AxirZ iivBpk"
            >
              Test task 1
            </p>
          </td>
          <td
            class="sc-AxiKw yXMWH"
            data-testid="table-description-cell-0"
          >
            <span
              class="sc-AxjAm fcxugk"
              data-testid="collapse-button-task-0"
            >
              ➡
            </span>
          </td>
          <td>
            <span
              aria-label="Cross"
              aria-labelledby="Cross"
              data-testid="remove-task-0-button"
              role="img"
              style="cursor: pointer;"
            >
              ❌
            </span>
            <span
              aria-label="Cross"
              data-testid="edit-task-0-button"
              role="img"
              style="margin-left: 0.5em; cursor: pointer;"
            >
              ✏
            </span>
          </td>
        </tr>
        <tr
          class="sc-AxhCb gnHiQZ"
          color="purple"
          data-testid="table-row-1"
        >
          <td
            data-testid="table-id-cell-1"
          >
            <p
              class="sc-AxirZ iivBpk"
            >
              1
            </p>
          </td>
          <td
            data-testid="table-title-cell-1"
          >
            <p
              class="sc-AxirZ iivBpk"
            >
              Test task 2
            </p>
          </td>
          <td
            class="sc-AxiKw yXMWH"
            data-testid="table-description-cell-1"
          />
          <td>
            <span
              aria-label="Cross"
              aria-labelledby="Cross"
              data-testid="remove-task-1-button"
              role="img"
              style="cursor: pointer;"
            >
              ❌
            </span>
            <span
              aria-label="Cross"
              data-testid="edit-task-1-button"
              role="img"
              style="margin-left: 0.5em; cursor: pointer;"
            >
              ✏
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
